numpy/core/src/multiarray/nditer_api.c

```diff
@@ -2830,4 +2789,23 @@ npyiter_checkreducesize(NpyIter *iter, npy_intp count,
     }
     return count * (*reduce_innersize);
 }
+
+NPY_NO_EXPORT npy_bool
+npyiter_has_writeback(NpyIter *iter)
```

Comment:
Instead of this function, which traverses the flags a second time - can you just add an extra return value from `NpyIter_Dealloc`?

Perhaps most simple would be to rename it to

```
// internal helper
NpyIter_Dealloc_int(NpyIter *iter, npy_bool *did_writeback) {
    // as before, but set did_writeback
}
/* NUMPY_API */
NpyIter_Dealloc(NpyIter *iter) {
    npy_bool did_writeback;
    return NpyIter_Dealloc_int(iter, &did_writeback);
}
```

Or perhaps even `NpyIter_Dealloc_int(iter, warn_if_writeback_needed=True)`

Reply:
I had considered that, but this API seems cleaner to me. Only the rarely used python `nditer` needs the warning, where the C `NpyIter` is used very frequently. The more commonly used path should be as clean as possible. A modified `NpyIter_Dealloc` would go through a second function call and would contain extra logic even for the fast path, where for `nditer` the extra call and flag traversal should be a small part of the GC cleanup when tp_dealloc is called.

We could revisit this idea in the future when the dust has settled around `nditer` use, it is an internal refactoring of the code.

Reply:
My fear with the current approach is that it separates the check of whether cleanup is needed from the actual code doing the cleanup.

> a second function call

It wouldn't surprise me if the compiler just inlined the outer function - assuming that we're using the internal function definition, rather than the external function pointer.

> We could revisit this idea in the future 

True, this PR as is is already a nice improvement